fmt/src/formatter.rs

```diff
@@ -919,6 +967,32 @@ impl<'a, W: Write> Formatter<'a, W> {
         Ok(indented)
     }
 
+    /// Add a function context around a procedure and revert the context at the end of the procedure
+    /// regardless of the response
+    fn with_function_context(
+        &mut self,
+        context: FunctionDefinition,
+        mut fun: impl FnMut(&mut Self) -> Result<()>,
+    ) -> Result<()> {
+        self.context.function = Some(context);
```

Comment:
heh, did a similar thing for parent pointer for `forge doc` with the exception of preserving the outer parent.